File: lib/nn.js
// Other techniques for learning

class ActivationFunction {
  constructor(func, dfunc) {
    this.func = func;
    this.dfunc = dfunc;
  }
}

let sigmoid = new ActivationFunction(
  x => 1 / (1 + Math.exp(-x)),
  y => y * (1 - y)
);

let tanh = new ActivationFunction(
  x => Math.tanh(x),
  y => 1 - (y * y)
);


class NeuralNetwork {
  /*
  * if first argument is a NeuralNetwork the constructor clones it
  * USAGE: cloned_nn = new NeuralNetwork(to_clone_nn);
  */
  constructor(in_nodes, hid_nodes, out_nodes) {
    if (in_nodes instanceof NeuralNetwork) {
      let a = in_nodes;
      this.input_nodes = a.input_nodes;
      this.hidden_nodes = a.hidden_nodes;
      this.output_nodes = a.output_nodes;

      this.weights_ih = a.weights_ih.copy();
      this.weights_ho = a.weights_ho.copy();

      this.bias_h = a.bias_h.copy();
      this.bias_o = a.bias_o.copy();
    } else {
      this.input_nodes = in_nodes;
      this.hidden_nodes = hid_nodes;
      this.output_nodes = out_nodes;

      this.weights_ih = new Matrix(this.hidden_nodes, this.input_nodes);
      this.weights_ho = new Matrix(this.output_nodes, this.hidden_nodes);
      this.weights_ih.randomize();
      this.weights_ho.randomize();

      this.bias_h = new Matrix(this.hidden_nodes, 1);
      this.bias_o = new Matrix(this.output_nodes, 1);
      this.bias_h.randomize();
      this.bias_o.randomize();
    }

    // TODO: copy these as well
    this.setLearningRate();
    this.setActivationFunction();


  }

  predict(input_array) {

    // Generating the Hidden Outputs
    let inputs = Matrix.fromArray(input_array);
    let hidden = Matrix.multiply(this.weights_ih, inputs);
    hidden.add(this.bias_h);
    // activation function!
    hidden.map(this.activation_function.func);

    // Generating the output's output!
    let output = Matrix.multiply(this.weights_ho, hidden);
    output.add(this.bias_o);
    output.map(this.activation_function.func);

    // Sending back to the caller!
    return output.toArray();
  }

  setLearningRate(learning_rate = 0.1) {
    this.learning_rate = learning_rate;
  }

  setActivationFunction(func = sigmoid) {
    this.activation_function = func;
  }

  train(input_array, target_array) {
    // Generating the Hidden Outputs
    let inputs = Matrix.fromArray(input_array);
    let hidden = Matrix.multiply(this.weights_ih, inputs);
    hidden.add(this.bias_h);
    // activation function!
    hidden.map(this.activation_function.func);

    // Generating the output's output!
    let outputs = Matrix.multiply(this.weights_ho, hidden);
    outputs.add(this.bias_o);
    outputs.map(this.activation_function.func);

    // Convert array to matrix object
    let targets = Matrix.fromArray(target_array);

    // Calculate the error
    // ERROR = TARGETS - OUTPUTS
    let output_errors = Matrix.subtract(targets, outputs);

    // let gradient = outputs * (1 - outputs);
    // Calculate gradient
    let gradients = Matrix.map(outputs, this.activation_function.dfunc);
    gradients.multiply(output_errors);
    gradients.multiply(this.learning_rate);


    // Calculate deltas
    let hidden_T = Matrix.transpose(hidden);
    let weight_ho_deltas = Matrix.multiply(gradients, hidden_T);

    // Adjust the weights by deltas
    this.weights_ho.add(weight_ho_deltas);
    // Adjust the bias by its deltas (which is just the gradients)
    this.bias_o.add(gradients);

    // Calculate the hidden layer errors
    let who_t = Matrix.transpose(this.weights_ho);
    let hidden_errors = Matrix.multiply(who_t, output_errors);

    // Calculate hidden gradient
    let hidden_gradient = Matrix.map(hidden, this.activation_function.dfunc);
    hidden_gradient.multiply(hidden_errors);
    hidden_gradient.multiply(this.learning_rate);

    // Calcuate input->hidden deltas
    let inputs_T = Matrix.transpose(inputs);
    let weight_ih_deltas = Matrix.multiply(hidden_gradient, inputs_T);

    this.weights_ih.add(weight_ih_deltas);
    // Adjust the bias by its deltas (which is just the gradients)
    this.bias_h.add(hidden_gradient);

    // outputs.print();
    // targets.print();
    // error.print();
  }

  serialize() {
    return JSON.stringify(this);
  }

  static deserialize(data) {
    if (typeof data == 'string') {
      data = JSON.parse(data);
    }
    let nn = new NeuralNetwork(data.input_nodes, data.hidden_nodes, data.output_nodes);
    nn.weights_ih = Matrix.deserialize(data.weights_ih);
    nn.weights_ho = Matrix.deserialize(data.weights_ho);
    nn.bias_h = Matrix.deserialize(data.bias_h);
    nn.bias_o = Matrix.deserialize(data.bias_o);
    nn.learning_rate = data.learning_rate;
    return nn;
  }


  // Adding function for neuro-evolution
  copy() {
    return new NeuralNetwork(this);
  }

  // Accept an arbitrary function for mutation
  mutate(func) {
    this.weights_ih.map(func);
    this.weights_ho.map(func);
    this.bias_h.map(func);
    this.bias_o.map(func);
  }

  visualize(x, y, isBlack = false, context) {
    if (context) {

      let input_points = [];
      let hidden_points = [];
      let output_points = [];

      for (let i = 0; i < this.input_nodes; i++) {
        let pointY = y + i * 20 - this.input_nodes * 10;
        context.beginPath();

        context.fillStyle = "white";
        context.strokeStyle = "white";

        context.ellipse(x, pointY, 5, 5, Math.PI / 4, 0, 2 * Math.PI);
        
        input_points.push(pointY);

        context.stroke();
        context.fill();
      }

      for (let i = 0; i < this.hidden_nodes; i++) {
        let pointY = y + i * 20 - this.hidden_nodes * 10;
        context.beginPath();

        context.fillStyle = "white";
        context.strokeStyle = "white";

        context.ellipse(x + 75, pointY, 5, 5, Math.PI / 4, 0, 2 * Math.PI);
        
        hidden_points.push(pointY);

        context.stroke();
        context.fill();
      }

      for (let i = 0; i < this.output_nodes; i++) {
        let pointY = y + i * 20 - this.output_nodes * 10;
        context.beginPath();

        context.fillStyle = "white";
        context.strokeStyle = "white";

        context.ellipse(x + 150, pointY, 5, 5, Math.PI / 4, 0, 2 * Math.PI);
        output_points.push(pointY);

        context.stroke();
        context.fill();
      }

      context.lineWidth = 1;

      for (let i = 0; i < this.input_nodes; i++) {
        for (let j = 0; j < this.hidden_nodes; j++) {
          context.beginPath();

          if (this.weights_ih.data[i][j] < 0) {
            context.strokeStyle = "red";

            context.globalAlpha = -this.weights_ih.data[i][j];
          } else {
            context.strokeStyle = "blue";

            context.globalAlpha = this.weights_ih.data[i][j];
          }

          context.moveTo(x, input_points[i]);
          context.lineTo(x + 75, hidden_points[j]);

          context.stroke();
        }
      }

      for (let i = 0; i < this.output_nodes; i++) {
        for (let j = 0; j < this.hidden_nodes; j++) {
          context.beginPath();

          if (this.weights_ho.data[i][j] < 0) {
            context.strokeStyle = "red";

            context.globalAlpha = -this.weights_ho.data[i][j];
          } else {
            context.strokeStyle = "blue";

            context.globalAlpha = this.weights_ho.data[i][j];
          }

          context.moveTo(x + 75, hidden_points[j]);
          context.lineTo(x + 150, output_points[i]);

          context.stroke();
        }
      }

      context.globalAlpha = 1;

    } else {
      let input_points = [];
      let hidden_points = [];
      let output_points = [];

      let colorMode;

      if (isBlack) {
        colorMode = color(0);
      } else if (!isBlack) {
        colorMode = color(255);
      }

      strokeWeight(1);
      fill(colorMode);
      stroke(colorMode);

      for (let i = 0; i < this.input_nodes; i++) {
        let pointY = y + i * 20 - this.input_nodes * 10;
        ellipse(x, pointY, 10, 10);
        
        input_points.push(pointY);
      }

      for (let i = 0; i < this.hidden_nodes; i++) {
        let pointY = y + i * 20 - this.hidden_nodes * 10;
        ellipse(x + 75, pointY, 10, 10);
        hidden_points.push(pointY);
      }

      for (let i = 0; i < this.output_nodes; i++) {
        let pointY = y + i * 20 - this.output_nodes * 10;
        ellipse(x + 150, pointY, 10, 10);
        output_points.push(pointY);
      }

      strokeWeight(1);
      let lineColor;

      for (let i = 0; i < this.input_nodes; i++) {
        for (let j = 0; j < this.hidden_nodes; j++) {
          if (this.weights_ih.data[i][j] < 0) {
            lineColor = color(255,0,0);
            lineColor.setAlpha(-this.weights_ih.data[i][j] * 255);
          } else {
            lineColor = color(0, 0, 255);
            lineColor.setAlpha(this.weights_ih.data[i][j] * 255);
          }
          stroke(lineColor);
          line(x, input_points[i], x + 75, hidden_points[j]);
        }
      }

      for (let i = 0; i < this.output_nodes; i++) {
        for (let j = 0; j < this.hidden_nodes; j++) {
          if (this.weights_ho.data[i][j] < 0) {
            lineColor = color(255,0,0);
            lineColor.setAlpha(-this.weights_ho.data[i][j] * 255);
          } else {
            lineColor = color(0, 0, 255);
            lineColor.setAlpha(this.weights_ho.data[i][j] * 255);
          }
          stroke(lineColor);
          line(x + 75, hidden_points[j], x + 150, output_points[i]);
        }
      }
    }
  }
}

File: python/nn.py
# Alen
# @000alen

from json import loads, dumps
from math import exp, tanh as mTanh
from random import uniform


def Array(rows, fill=None):
	array = []
	for x in range(rows):
		array.append(fill)
	return array


class Matrix:
	def __init__(self, rows, cols):
		self.map = self.__map__
		self.multiply = self.__multiply__

		self.rows = rows
		self.cols = cols
		self.data = map(lambda i: Array(self.cols, 0), Array(self.rows))

	@staticmethod
	def fromArray(arr):
		return Matrix(len(arr), 1).map(lambda e, i, j: arr[i])

	@staticmethod
	def subtract(a, b):
		if not a.rows == b.rows or not a.cols == b.cols:
			raise Exception
		return Matrix(a.rows, a.cols).map(lambda _, i, j: a.data[i][j] - b.data[i][j])

	def toArray(self):
		arr = []
		for i in range(self.rows):
			for j in range(self.cols):
				arr.append(self.data[i][j])
		return arr

	def randomize(self):
		return self.map(lambda e, i, j: (uniform(0, 1) * 2) + 1)

	def add(self, n):
		if n.__class__ == self.__class__:
			if not self.rows == n.rows or not self.cols == n.cols:
				raise Exception
			else:
				return self.map(lambda e, i, j: e + n.data[i][j])
		else:
			return self.map(lambda e, i, j: e + n)

	@staticmethod
	def transpose(matrix):
		return Matrix(matrix.cols, matrix.rows).map(lambda _, i, j: matrix.data[j][i])

	@staticmethod
	def multiply(a, b):
		def __lambda__(e, i, j):
			sum = 0
			for k in range(a.cols):
				sum += a.data[i][k] * b.data[k][j]
			return sum

		if not a.cols == b.rows:
			raise Exception
		return Matrix(a.rows, b.cols).map(__lambda__)

	def __multiply__(self, n):
		if n.__class__ == self.__class__:
			if not self.rows == n.rows or not self.cols == n.cols:
				raise Exception
			return self.map(lambda e, i, j: e * n.data[i][j])
		else:
			return self.map(lambda e, i, j: e * n)

	@staticmethod
	def map(matrix, func):
		return Matrix(matrix.rows, matrix.cols).map(lambda e, i, j: func(matrix.data[i][j], i, j))

	def __map__(self, func):
		for i in range(self.rows):
			for j in range(self.cols):
				val = self.data[i][j]
				self.data[i][j] = func(val, i, j)
		return self

	def print_(self):
		print str(self.data).replace("],", "],\n")
		return self

	def serialize(self):
		return dumps(self.data)

	@staticmethod
	def deserialize(data):
		if type(data) is type(str()):
			data = loads(data)
		matrix = Matrix(data.rows, data.cols)
		matrix.data = data.data
		return matrix


class ActivationFunction:
	def __init__(self, func, dfunc):
		self.func = func
		self.dfunc = dfunc


class NeuronalNetwork:
	def __init__(self, input_nodes, hidden_nodes, output_nodes):
		self.input_nodes = input_nodes
		self.hidden_nodes = hidden_nodes
		self.output_nodes = output_nodes

		self.weights_ih = Matrix(self.hidden_nodes, self.input_nodes)
		self.weights_ho = Matrix(self.output_nodes, self.hidden_nodes)
		self.weights_ih.randomize()
		self.weights_ho.randomize()

		self.bias_h = Matrix(self.hidden_nodes, 1)
		self.bias_o = Matrix(self.output_nodes, 1)
		self.bias_h.randomize()
		self.bias_o.randomize()
		self.setLearningRate()

		self.setActivationFunction()

	def predict(self, input_array):
		inputs = Matrix.fromArray(input_array)
		hidden = Matrix.multiply(self.weights_ih, inputs)
		hidden.add(self.bias_h)

		hidden.map(self.activation_function.func)

		output = Matrix.multiply(self.weights_ho, hidden)
		output.add(self.bias_o)
		output.map(self.activation_function.func)

		return output.toArray()

	def setLearningRate(self, learning_rate = 0.1):
		self.learning_rate = learning_rate

	def setActivationFunction(self, func = sigmoid):
		self.activation_function = func

	def train(self, input_array, target_array):
		inputs = Matrix.fromArray(input_array)

		hidden = Matrix.multiply(self.weights_ih, inputs)
		hidden.add(self.bias_h)
		hidden.map(self.activation_function.func)

		outputs = Matrix.multiply(self.weights_ho, hidden)
		outputs.add(self.bias_o)
		outputs.map(self.activation_function.func)

		targets = Matrix.fromArray(target_array)

		output_errors = Matrix.subtract(targets, outputs)

		gradients = Matrix.map(outputs, self.activation_function.dfunc)
		gradients.multiply(output_errors)
		gradients.multiply(self.learning_rate)


		hidden_T = Matrix.transpose(hidden)
		weight_ho_deltas = Matrix.multiply(gradients, hidden_T)

		self.weights_ho.add(weight_ho_deltas)
		self.bias_o.add(gradients)

		who_t = Matrix.transpose(self.weights_ho)
		hidden_errors = Matrix.multiply(who_t, output_errors)

		hidden_gradient = Matrix.map(hidden, self.activation_function.dfunc)
		hidden_gradient.multiply(hidden_errors)
		hidden_gradient.multiply(self.learning_rate)

		inputs_T = Matrix.transpose(inputs)
		weight_ih_deltas = Matrix.multiply(hidden_gradient, inputs_T)

		self.weights_ih.add(weight_ih_deltas)
		self.bias_h.add(hidden_gradient)

	def serialize(self):
		data = {
			"input_nodes": self.input_nodes,
			"hidden_nodes": self.hidden_nodes,
			"output_nodes": self.output_nodes,
			"weights_ih": self.weights_ih,
			"weights_ho": self.weights_ho,
			"bias_h": self.bias_h,
			"bias_o": self.bias_o,
			"learning_rate": self.learning_rate
		}
		return dumps(data)

	@staticmethod
	def deserialize(data):
		if type(data) == type(str()):		
			data = loads(data)
		nn = NeuronalNetwork(data["input_nodes"], data["hidden_nodes"], data["output_nodes"])
		nn.weights_ih = Matrix.deserialize(data["weights_ih"])
		nn.weights_ho = Matrix.deserialize(data["weights_ho"])
		nn.bias_h = Matrix.deserialize(data["bias_h"])
		nn.bias_o = Matrix.deserialize(data["bias_o"])
		nn.learning_rate = data["learning_rate"]
		return nn

sigmoid = ActivationFunction(
	lambda x, i, j: 1 / (1 + exp(-x)),
	lambda y, i, j: y * (1 - y)
)

tanh = ActivationFunction(
	lambda x, i, j: mTanh(x),
	lambda y, i, j: 1 - (y ** 2)
)
